src/main/scala/analysis/solvers/DSAUnionFindSolver.scala

```diff
@@ -0,0 +1,72 @@
+package analysis.solvers
+
+import analysis.DSN
+
+import scala.collection.mutable
+
+class DSAUnionFindSolver extends UnionFindSolver[UniTerm] {
+
+
+  val parent = mutable.Map[DSAUniTerm, DSAUniTerm]()
+  val offsets = mutable.Map[DSAUniTerm, BigInt]()
+
+  override def unify(t1: Term[UniTerm], t2: Term[UniTerm]): Unit =
+    unify(t1.asInstanceOf[DSAUniTerm], t2.asInstanceOf[DSAUniTerm], 0)
+
+  // offset is the offset at which
+  def unify(t1: DSAUniTerm, t2: DSAUniTerm, offset: BigInt): Unit = {
+    mkSet(t1)
+    mkSet(t2)
+    val rep1 = findWithOffset(t1)._1
+    val rep2 = findWithOffset(t2)._1
+
+    if (rep1 == rep2) return
+
+      (rep1, rep2) match {
+        case (t1: DSAUniTerm, t2: DSAUniTerm) =>
+          mkUnion(t1, t2, offset)
+        case (x, y) =>
+          throw new UnificationFailure(s"Cannot unify $t1 and $t2 (with representatives $x and $y)")
+      }
```

Comment:
this can just be mkUnion(t1, t2, offset) without the match statement since t1 and t2 are always of type DSAUniTerm